mobile-browser-based-version/src/core/communication/decentralized_client.ts

```diff
@@ -0,0 +1,308 @@
+import { List, Map, Seq, Set } from 'immutable'
+import msgpack from 'msgpack-lite'
+import SimplePeer from 'simple-peer'
+import tf from '@tensorflow/tfjs'
+
+import {
+  deserializeWeights,
+  isSerializedVariable,
+  SerializedVariable,
+  serializeWeights
+} from './serialization'
+import { Client } from './client'
+import { Task } from '@/core/task/base/task'
+import { TrainingInformant } from '@/core/training/training_informant'
+
+type Weights = tf.LayerVariable[]
+type PeerMessage = { epoch: number, weights: SerializedVariable[] }
+
+// TODO take it from the server sources
+type PeerID = number
+type EncodedSignal = Uint8Array
+type ServerOpeningMessage = PeerID[]
+type ServerPeerMessage = [PeerID, EncodedSignal]
+
+function isPeerMessage (data: unknown): data is PeerMessage {
```

Comment:
Oh! This is cool, I didn't know the return type could be a statement like that! (I guess implicitly it's a boolean? but we show explicitly what we are checking?)

Reply:
> Oh! This is cool, I didn't know the return type could be a statement like that!

Yep, that's a very nice feature of TS, look at https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates for more information :)

> I guess implicitly it's a boolean?

Indeed, it's common for types in JS to be checked this way (`Array.isArray`, `Buffer.isBuffer`, ...).

> but we show explicitly what we are checking?

This way, the compiler knows that when branching with this function, the object is of this type (± implementing the `is` function correctly)

```
if (!isPeerMessage(msg)) {
  throw new Error('msg is not valid')
}

// here `msg` is of type `PeerMessage`
```